### ISSUE TITLE
Partial synchronization, synchronization order and minimum free disk space parameters

### ISSUE THREAD

**Issue by hiqua** [syncthing/syncthing#5948]
## Summary

In case of partial synchronization, the synchronization order and the minimum free disk space parameters are not properly taken into account.

## Situation

I have a folder which I want to synchronize in alphabetic order, with a minimum free disk space of 8% (for **this folder**, not the global free disk space parameter of the database) on a device D, synchronized from another device S containing the full folder. I expect this synchronization to be sometimes partial, meaning that the folder can be bigger than there is free space on D.

## Expected behavior

Ideally, the partial sync would still take the two parameters into account, meaning that, by the time it gets stuck (because there's no more memory available):
1. the files on D are the first in alphabetic (consecutive) order on S
2. there's still at least 8% free disk space on D

I would expect Syncthing to compute which files it can download given the current free disk space, and to stop when the minimum free disk space is reached. I would also expect Syncthing to favor one (potentially partial) file over several partial files . At least, I would also expect that no file is partially downloaded if the free space is less than the given parameter.
## Actual behavior

Currently, both conditions are violated:
1. fully downloaded files are not in alphabetic consecutive order on S (some files are partially downloaded, i.e., there are multiple -- 4 last time I tried -- .syncthing.$FILENAME files in the folder)
2. there's very little space left (less than 1%).

## Versions:
* S runs on 1.0.0, Debian Buster
* D on 1.2.0, Android 9 (using the `official` app, not the fork)

I didn't see anything in the changelog of 1.2.1 that could help mitigate this issue.

Edit: updated to add details asked in comments

**Comment by AudriusButkevicius** [comment]
I suggest you start this with a thread on the forum to debug the issue, as there is not much to work from now. No logs, etc. Do you have evidence that syncthing turned the 8% into 1% oppose to some other app? Also, there are two min space settings, one for the files, one for the database, are both set at 8%

**Comment by imsodin** [comment]
1. Indeed we do not abort synchronization when the free disk space requirement is violated by any one of the files to be synced. And in my opinion this is a good behaviour, as it means that as many files as possible are synced within the limits set by the user (imagine one huge file and many tiny ones). Maybe there is a case to be made for aborting synchronization if pull order is not random, but I am not entirely convinced this would be beneficial for a wider audience. As Audrius said, please take that discussion to the forum and we can revisit once a consensus has been reached.

2. I think that can happen if two consecutive files are pulled, where each individually do not violate the free disk space, but both together do. then if the first passes the check, and there's a race between doing the check for the second file and allocating the first file on disk. Or it's not Syncthing's fault at all. So if #5949 will not fix your issue, please report back with steps to reproduce or logs that show the problem.

**Comment by hiqua** [comment]
I've added more information to the original post to take your questions into account.

1. is really only about the case of deterministic pull order, indeed in the random case I don't think one should expect anything in particular (although a full file is preferable to two partial files). The reason I submitted this issue is that I think partial synchronizations is one of the two cases where this parameter is useful (the other one being slow connections).

2. I had 4 partial files last time I tried. I looked for some *advanced* parameter controlling how many files are synced in parallel, as setting this to 1 would be a potential workaround (obviously there are better solutions), but there doesn't seem to be any.

I'll ask in the forum what people expect when they use the pull order parameter and I'll provide exact steps to reproduce this issue within the next 3 weeks. If you release the fix for #5949 by then I'll try to take it into account to make them more relevant. But in short, I think it's enough to set the free disk space parameter to `10%`, have files `a`, `b`, `c`, etc. with decreasing sizes and totaling something like 20% of the free disk space and being less than 10% individually, and set the pull order to `alphabetic`. I assume that the sizes also have to be different enough so that when a batch of files is synchronized, some of them are downloaded noticeably faster than others. Given that it's more or less a concurrency issue, I also don't expect it to necessarily happen all the time, although I've had it every time in my case.

**Comment by AudriusButkevicius** [comment]
I disagree that we should terminate sync if the next file in the pull order does not fit on disk. I think we should sync as many files as we can, regardless of the order. The order is advisory at best anyway.

Also, my point is that folder and db limits are not connected. Sure, the folder stopped downloading stuff, but the database kept growing breaking the 8%, if they happen to colocated on the same storage.

**Comment by AudriusButkevicius** [comment]
I am not sure if there is anything left to fix here. The PR moves the check deeper preveting a potential race, but other than that it's not clear what needs fixing here or how to reproduce it.

**Comment by hiqua** [comment]
Given that there's already a fix, I'll wait to have it on my machines (which might take a while) to test if I can still reproduce the bug. I'll reopen this issue if that's the case.

I finally got your point regarding db size vs folder size, thanks for the explanation (I had assumed that the db size was negligible, but I didn't check).

I think it does make sense to synchronize one partial file, as long as we don't end up in the situation where we have *several* partial files while we could instead of a full big one. Of course it's fine to have one partial file which current size is bigger than the total size of other files (e.g., they could have been added to the folder later preventing any meaningful ordering).
